### DISCUSSION TITLE
Best practice when working with large Many relations

### DISCUSSION THREAD

**Discussion by chrisbenincasa** [mikro-orm/mikro-orm#5627] [Q&A] [answered]
I'm wondering what the best practice is for working with large 1:M and M:M relationships with mikro-orm. I have a scenario where relations are frequently cleared (`removeAll`) and then reset entirely. For "huge" relations, this can cause errors like: 
```
 - Expression tree is too large (maximum depth 1000)
```

when removing all of the relations at once. The query generated is an enormous `delete` with chained `or`s. Is there any way to chunk this? Or are there better ways to approach the operation?

For our use-case, this operation is generally followed by a large collection `set` for the incoming IDs, which potentially suffers from the same issue.

**Comment by B4nan** [comment]
>  - Expression tree is too large (maximum depth 1000)

First time hearing about this, can you provide a reproduction? Maybe we could handle this automatically, we already deal with limits for batching for insert and update queries for a very similar reason.

**Comment by chrisbenincasa** [comment]
Sure! https://github.com/chrisbenincasa/reproduction-populate-hint/tree/max-depth

I did some digging on the error itself, too, and it looks like an internal limit in sqlite:

https://www.sqlite.org/limits.html

> The SQLITE_MAX_EXPR_DEPTH parameter determines the maximum expression tree depth. If the value is 0, then no limit is enforced. The current implementation has a default value of 1000.

**Reply by B4nan** [reply]
Thanks, will see what I can do. Btw if [this test](https://github.com/chrisbenincasa/reproduction-populate-hint/blob/max-depth/src/example.test.ts#L87-L90) represents what you are doing, I would suggest you skip the intermediate flush call, it makes no sense to do it this way. Calling `Collection.set` replaces the values - if its loaded, it will make a diff and remove/insert what is needed, if its not loaded, it will wipe the whole collection first and only add new ones. So you can as well skip the `removeAll` call. Lastly, you don't need to persist managed entities, that's another no-op.

```diff
-a.b.removeAll();
-await em.persistAndFlush(a);
-a.b.set(newBs);
-await em.persistAndFlush(a);
+a.b.set(newBs);
+await em.flush();
```

**Reply by B4nan** [reply]
Based on a quick look, this can be easily handled internally by batching the delete calls too. The batch size will be configurable (we already have an option for that), so if your query would end up too heavy, you can adjust the limits based on your environment (or bump the sqlite limit instead).

**Reply by chrisbenincasa** [reply]
Thanks! Yea I think our original code did something like that until I realized (from digging through the code here) that set would do the difference like you state (I don't think that is mentioned in the docs, but correct me if I'm wrong and I missed it). Either way, I eventually had to move to chunked add/removes because of the issue here. Either way, I'm still working on my mental model of everything here! 

Good to know that this is potentially an easy fix, however!

**Reply by B4nan** [reply]
Actually, looking at the code, we are already batching the deletes, we just don't batch the collection updates which are handled separately.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/unit-of-work/ChangeSetPersister.ts#L73-L76

**Reply by B4nan** [reply, marked as answer]
Ok, it was pretty simple: d97979b236b8a7e40f64848e88602c531b8ebb84

You can use the `batchSize` option if the default (300 items per batch) wouldn't work for you.

**Reply by B4nan** [reply]
I was trying out the other case and noticed it was much slower, literally 800ms slower, doing exactly the same. And indeed, the diffing was very slow, managed to fix that here: f46e7c86e29727b57f3220901ac5b14a6f1719c1

**Reply by chrisbenincasa** [reply]
This is great stuff - appreciate the super quick turnaround. Perf improvement there is great too!
